pkg/storage/replica.go

```diff
@@ -900,11 +900,25 @@ func (r *Replica) updateProposalQuotaRaftMuLocked(
 
 	// We're still the leader.
 
+	// We need to check if the replica is being destroyed and if so, close the
+	// quota pool and unblock ongoing quota acquisition goroutines (if any).
+	if r.mu.destroyed != nil {
```

Comment:
Strikes me as odd that we need to do this here, and there is still a potential problem here that leaves clients hanging if the replica gets destroyed but this code path is never taken. Having taken a rough look at the ways in which replicas can get destroyed unfortunately convinced me that you shouldn't be burdened with addressing that as well - it's a bit of a mess.